server/src/main/java/io/druid/client/selector/ServerSelector.java

```diff
@@ -19,23 +19,27 @@
 
 package io.druid.client.selector;
 
+import io.druid.java.util.common.RE;
 import io.druid.server.coordination.DruidServerMetadata;
+import io.druid.server.coordination.ServerType;
 import io.druid.timeline.DataSegment;
 import it.unimi.dsi.fastutil.ints.Int2ObjectRBTreeMap;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.stream.Collectors;
 
 /**
  */
 public class ServerSelector implements DiscoverySelector<QueryableDruidServer>
 {
 
-  private final Int2ObjectRBTreeMap<Set<QueryableDruidServer>> servers;
+  private final Int2ObjectRBTreeMap<Set<QueryableDruidServer>> historicalServers;
+
+  private final Int2ObjectRBTreeMap<Set<QueryableDruidServer>> realtimeServers;
```

Comment:
can't we just use a sorted map for `servers` that puts historicals before peon and that alone would get us the preference to historicals ?

Reply:
The won't exactly work because if we pass this sorted map (hist + realtime) to the strategies, we will have to make the strategies understand what realtime, historical etc. For example in the RandomServerSelectorStrategy, just picks servers at random and will not know how to interpret realtime, hist etc.